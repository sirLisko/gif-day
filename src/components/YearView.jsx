import React, { useState } from "react";
import PropTypes from "prop-types";
import { getDaysInMonth } from "date-fns";
import times from "lodash.times";
import styled from "@emotion/styled/macro";

import theme from "styles/theme";
import GifTile from "components/GifTile";

const StyledTable = styled.table`
  margin: 0 auto;
  ${theme.mq.max.medium} {
    tbody {
      display: flex;
    }
  }
  td {
    margin: 0;
    padding: 0;
  }
`;

const StyledMonth = styled.tr`
  ${theme.mq.max.medium} {
    display: flex;
    flex-direction: column;
  }
`;

const StyledDay = styled.button(
  (props) => `
  border: 1px solid ${theme.colors.gray_light};
  color: ${theme.colors.gray_light};
  border-radius: 0.5rem;
  background: none;
  text-align: center;
  padding: 0;
  margin: 0;
  overflow: hidden;
  height: 8vw;
  width: 8vw;
  font-size: 0.25rem;
  ${theme.mq.small} {
    font-size: 0.5rem;
  }
  ${theme.mq.medium} {
    height: 3vw;
    width: 3vw;
  }
  div {
    width: 100%;
    object-fit: cover;
    height: 100%;
  }
  img,
  video {
    width: 100%;
    height: 100%;
  }
  img {
    display: ${props.dynamic ? "block" : "none"};
  }
  video {
    display: ${props.dynamic ? "none" : "block"};
  }
  &:hover img {
    display: none;
  }
  &:hover video {
    display: block;
  }
  &:hover {
    color: #000;
    border-color: #000;
  }
`
);

export const YearView = ({ dailyGifs, onSelectedDay }) => {
  const [thisYear] = useState(new Date().getFullYear());
  const dayOfTheMonth = times(12).map((month) =>
    getDaysInMonth(new Date(thisYear, month))
  );

  return (
    <StyledTable>
      <tbody>
        {dayOfTheMonth.map((days, monthIndex) => (
          <StyledMonth key={monthIndex}>
            {times(days).map((day) => {
              const index = `${day}-${monthIndex}`;
              return (
                <td key={index}>
                  <StyledDay onClick={() => onSelectedDay(index)} role="button">
                    {dailyGifs[index] ? (
                      <GifTile gifObj={dailyGifs[index].gif} />
                    ) : (
                      `${day + 1} / ${monthIndex + 1}`
                    )}
                  </StyledDay>
                </td>
              );
            })}
          </StyledMonth>
        ))}
      </tbody>
    </StyledTable>
  );
};

YearView.propTypes = {
  dailyGifs: PropTypes.shape().isRequired,
  onSelectedDay: PropTypes.func.isRequired,
};

export default YearView;
