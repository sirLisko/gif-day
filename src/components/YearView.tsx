import React, { useState } from "react";
import { getDaysInMonth } from "date-fns";
import times from "lodash.times";

import theme from "styles/theme";
import GifTile from "components/GifTile";
import { DailyGifs } from "types";

interface Props {
  dailyGifs: DailyGifs;
  onSelectedDay: (dayIndex: string) => void;
}

export const YearView = ({ dailyGifs, onSelectedDay }: Props) => {
  const [thisYear] = useState(new Date().getFullYear());
  const dayOfTheMonth = times(12).map((month) =>
    getDaysInMonth(new Date(thisYear, month)),
  );

  return (
    <table>
      <tbody>
        {dayOfTheMonth.map((days, monthIndex) => (
          <tr key={monthIndex}>
            {times(days).map((day) => {
              const index = `${day}-${monthIndex}`;
              return (
                <td key={index}>
                  <button onClick={() => onSelectedDay(index)} role="button">
                    {dailyGifs[index] ? (
                      <GifTile gifObj={dailyGifs[index]} />
                    ) : (
                      `${day + 1} / ${monthIndex + 1}`
                    )}
                  </button>
                </td>
              );
            })}
          </tr>
        ))}
      </tbody>
      <style jsx>{`
        table {
          margin: 0 auto;
          ${theme.mq.max.medium} {
            tbody {
              display: flex;
            }
          }
          td {
            margin: 0;
            padding: 0;
          }
        }
        tr {
          ${theme.mq.max.medium} {
            display: flex;
            flex-direction: column;
          }
        }
        button {
          border: 1px solid ${theme.colors.gray_light};
          color: ${theme.colors.gray_light};
          border-radius: 0.5rem;
          background: none;
          text-align: center;
          padding: 0;
          margin: 0;
          overflow: hidden;
          height: 8vw;
          width: 8vw;
          font-size: 0.25rem;
          ${theme.mq.small} {
            font-size: 0.5rem;
          }
          ${theme.mq.medium} {
            height: 3vw;
            width: 3vw;
          }
          div {
            width: 100%;
            object-fit: cover;
            height: 100%;
          }
          img,
          video {
            width: 100%;
            height: 100%;
          }
          .dynamic img {
            display: block;
          }
          img {
            display: none;
          }
          .dynamic video {
            display: none;
          }
          video {
            display: block;
          }
          &:hover img {
            display: none;
          }
          &:hover video {
            display: block;
          }
          &:hover {
            color: #000;
            border-color: #000;
          }
        }
      `}</style>
    </table>
  );
};

export default YearView;
