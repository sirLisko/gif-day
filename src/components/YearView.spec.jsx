import React from 'react';
import { shallow } from 'enzyme';
import mockdate from 'mockdate';

import YearView from './YearView';

describe('YearView Component', () => {
  let wrapper;
  const props = {
    dailyGifs: {},
    onSelectedDay: jest.fn(),
  };

  beforeEach(() => {
    wrapper = shallow(<YearView {...props} />);
    props.onSelectedDay.mockReset();
  });

  it('should render properly', () => {
    wrapper = shallow(<YearView {...props} />);
    expect(wrapper.find('StyledMonth').length).toBe(12);
    expect(wrapper.find('StyledDay').first()).toMatchSnapshot();
  });

  it('should render the correct number of days if leap year', () => {
    jest.resetModules();
    mockdate.set('1/1/2019');
    let YearView = require('./YearView').default;
    wrapper = shallow(<YearView {...props} />);
    expect(wrapper.find('StyledDay').length).toBe(365);
  });

  it('should render the correct number of days if leap year', () => {
    jest.resetModules();
    mockdate.set('1/1/2000');
    let YearView = require('./YearView').default;
    wrapper = shallow(<YearView {...props} />);
    expect(wrapper.find('StyledDay').length).toBe(366);
  });

  it('should render properly with GIF', () => {
    wrapper.setProps({
      dailyGifs: {
        '0-0': {
          text: 'foo',
          src: 'bar',
        },
      },
    });
    expect(wrapper.find('StyledDay').first()).toMatchSnapshot();
  });

  it('should select the correct day', () => {
    wrapper.setProps({
      dailyGifs: {
        '0-0': {
          text: 'foo',
          src: 'bar',
        },
      },
    });
    wrapper
      .find('StyledDay')
      .first()
      .simulate('click');
    expect(props.onSelectedDay).toHaveBeenCalled();
    expect(props.onSelectedDay).toHaveBeenCalledWith('0-0');
  });
});
