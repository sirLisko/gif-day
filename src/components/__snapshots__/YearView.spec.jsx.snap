// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearView Component should render properly 1`] = `
<StyledDay
  onClick={[Function]}
>
  1 / 1
</StyledDay>
`;

exports[`YearView Component should render properly with GIF 1`] = `
<StyledDay
  onClick={[Function]}
>
  <GifTile
    gif="bar"
  />
</StyledDay>
`;
